src/Build/Graph/ProjectGraph.cs

```diff
@@ -317,10 +330,73 @@ public ProjectGraph(IEnumerable<ProjectGraphEntryPoint> entryPoints, ProjectColl
                 globalProperties);
-            _allParsedProjects.Add(configurationMetadata, graphNode);
+            _allParsedProjects[configurationMetadata] = graphNode;
             return graphNode;
         }
 
+        /// <summary>
+        /// Load a graph with root node at entryProjectFile
+        /// Maintain a queue of projects to be processed and evaluate projects in parallel
+        /// </summary>
+        private void LoadGraph(ConcurrentQueue<ConfigurationMetadata> projectsToEvaluate, ProjectCollection projectCollection, ConcurrentDictionary<ConfigurationMetadata, object> tasksInProgress)
+        {
+            var waitHandle = new AutoResetEvent(false);
+            while (projectsToEvaluate.Count != 0 || tasksInProgress.Count != 0)
+            {
+                ConfigurationMetadata projectToEvaluate;
+                if (projectsToEvaluate.Count != 0)
+                {
+                    projectToEvaluate = projectsToEvaluate.Dequeue();
+                    var task = new Task(() =>
+                    {
+                        ProjectGraphNode parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
+                        IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.ProjectInstance.GetItems(ProjectReferenceItemName);
+                        foreach (var projectReferenceToParse in projectReferenceItems)
+                        {
+                            if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                            {
+                                throw new InvalidOperationException(string.Format(
+                                    CultureInfo.InvariantCulture,
+                                    ResourceUtilities.GetResourceString(
+                                        "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
+                                    projectReferenceToParse.EvaluatedInclude,
+                                    parsedProject.ProjectInstance.FullPath));
+                            }
+
+                            string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
+                            PropertyDictionary<ProjectPropertyInstance> projectReferenceGlobalProperties = GetProjectReferenceGlobalProperties(projectReferenceToParse, projectToEvaluate.GlobalProperties);
+                            var projectReferenceConfigurationMetadata = new ConfigurationMetadata(projectReferenceFullPath, projectReferenceGlobalProperties);
+                            if (!tasksInProgress.ContainsKey(projectReferenceConfigurationMetadata))
+                            {
+                                if (!_allParsedProjects.ContainsKey(projectReferenceConfigurationMetadata))
+                                {
+                                    projectsToEvaluate.Enqueue(projectReferenceConfigurationMetadata);
```

Comment:
What if you set WaitHandle here instead? It would make the main thread loop spin new tasks faster right?

Reply:
It can't be set just here. It has to be set when the thread terminates as well so that the main loop exits after parsing all targets.

Reply:
Discussed offline: OK to set when either of the two collections gets changed. The extra sets don't seem to be harmful.